hoomd/md/tune/nlist_buffer.py

```diff
@@ -0,0 +1,322 @@
+
+        The tuner is considered tune when the specified solver returns ``True``
+        when solving twice consecutively. See `hoomd.tune` for more information
+        on tuning criteria.
+        """
+        return self._tuned > 1
+
+    @hoomd.logging.log
+    def max_tps(self):
+        """int: The maximum recorded TPS during tuning."""
+        return self._max_tps
+
+    @hoomd.logging.log
+    def best_buffer_size(self):
+        """float: The buffer size corresponding to ``max_tps`` during tuning."""
+        return self._best_buffer_size
+
+    @hoomd.logging.log
+    def last_tps(self):
+        """int: The last TPS computed for the tuner."""
+        return self._last_tps
+
+    @property
+    def final_bin(self):
+        """tuple[float, float]: Boundaries of grid search optimization.
+
+        Property is only available when a `hoomd.tune.GridOptimizer` is used,
+        and tuning is finished. Otherwise, the property is ``None``.
+        """
+        if not isinstance(self.solver, hoomd.tune.GridOptimizer):
+            return None
+        if self._tunable is None:
+            return None
+        return self.solver._final_bins.get(self._tunable, None)
+
+    def __getstate__(self):
+        state = copy.copy(self.__dict__)
+        for attr in self._skip_for_equality:
+            state.pop(attr, None)
+        return state
+
+
+class NeighborListBuffer(hoomd.tune.custom_tuner._InternalCustomTuner):
+    """Optimize neighbor list buffer size for maximum TPS.
+
+    Direct instantiation of this class requires a `hoomd.tune.Optimizer`
+    that determines how move sizes are updated. This class also provides class
+    methods to create a `NeighborListBuffer` tuner with built-in solvers; see
+    `NeighborListBuffer.with_grid` and
+    `NeighborListBuffer.with_gradient_descent`.
+
+    Args:
+        trigger (hoomd.trigger.Trigger): ``Trigger`` to determine when to run
+            the tuner.
+        nlist (hoomd.md.nlist.NeighborList): Neighbor list instance to tune.
+        solver (`hoomd.tune.solve.Optimizer`): A solver that tunes the
+            neighbor list buffer to maximize TPS.
+        maximum_buffer (float): The largest buffer value to allow.
+        minimum_buffer (`float`, optional): The smallest buffer value to allow
+            (defaults to 0).
+
+    Attributes:
+        trigger (hoomd.trigger.Trigger): ``Trigger`` to determine when to run
+            the tuner.
+        solver (`hoomd.tune.solve.Optimizer`): A solver that tunes the
+            neighbor list buffer to maximize TPS.
+        maximum_buffer (float): The largest buffer value to allow.
+        minimum_buffer (float): The smallest buffer value to allow.
+
+    Warning:
+        When using with a `hoomd.device.GPU` device, autotuning can mess with
+        convergence. Running the simulation before adding the tuner to allow the
+        autotuning to converge first results in better TPS optimization.
```

Comment:
```suggestion
        When using with a `hoomd.device.GPU` device, kernel launch parameter
        autotuning can prevent convergence. Run the simulation for 25,000 steps
        before adding the tuner to allow the autotuning to converge first
        results in better TPS optimization.
```
Add some specificity.